Neos.Neos/Classes/FrontendRouting/NodeUri/NodeUriBuilderFactory.php

```diff
@@ -0,0 +1,45 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Neos\Neos\FrontendRouting\NodeUri;
+
+use GuzzleHttp\Psr7\Uri;
+use Neos\Flow\Annotations as Flow;
+use Neos\Flow\Http\Helper\RequestInformationHelper;
+use Neos\Flow\Http\ServerRequestAttributes;
+use Neos\Flow\Mvc\Routing\Dto\RouteParameters;
+use Neos\Flow\Mvc\Routing\RouterInterface;
+use Psr\Http\Message\ServerRequestInterface;
+
+#[Flow\Scope('singleton')]
+final class NodeUriBuilderFactory
+{
+    /**
+     * The possibly configured Flow base URI, see {@see \Neos\Flow\Http\BaseUriProvider}
+     * @var string|null
+     */
+    #[Flow\InjectConfiguration(package: 'Neos.Flow', path: 'http.baseUri')]
+    protected $configuredBaseUri;
+
+    #[Flow\Inject]
+    protected RouterInterface $router;
+
+    /**
+     * @api
+     */
+    public function forRequest(ServerRequestInterface $request): NodeUriBuilder
```

Comment:
regarding this method and the mandatory use of `getHttpRequest` in this case:

```php
$this->nodeUriBuilderFactory
    ->forRequest($actionRequest->getHttpRequest())
```

I have written quite a monolog regarding how it could be a nice thought to use the whole action request in itself to pass along as context and make getHttpRequest nullable in CLI context. This might sound controversy but i think its a good idea to think about it before introducing something like `ControllerContext2.0` or `UriBuildingContext`:

Discussion: https://github.com/neos/flow-development-collection/issues/3354

Reply:
we decided for `forActionRequest` for now and might introduce a `forRouteDefaults` lateron